src/main/java/com/github/manolo8/darkbot/utils/login/LoginUtils.java

```diff
@@ -47,7 +50,57 @@ public static LoginData performUserLogin() {
         Popups.showMessageSync("Login", pane, panel::setDialog);
 
         LoginData loginData = panel.getResult();
-        if (loginData.getPreloaderUrl() == null || loginData.getParams() == null) System.exit(0);
+        if (loginData.getPreloaderUrl() == null || loginData.getParams() == null) {
+            System.out.println("Closed login panel, exited without logging in");
+            System.exit(0);
+        }
+        return loginData;
+    }
+
+    public static LoginData performAutoLogin(StartupParams params) {
+        Credentials credentials = loadCredentials();
+        String password = params.get(StartupParams.PropertyKey.PASSWORD);
+        Credentials.User user = null;
+        
+        if (password == null || password.isEmpty()) {
+            try {
+                char[] masterPassword = params.getMasterPassword();
+                if (masterPassword == null) {
+                    System.err.println("Both master-password and password are undefined, make sure to at least define one of these fields");
+                }
+                credentials.decrypt(masterPassword);
+            } catch (Exception e) {
+                System.err.println("Couldn't retreive logins, check your startup properties file");
+                e.printStackTrace();
+            }
+            
+            user = credentials.getUsers()
+                    .stream()
+                    .filter(usr -> usr.u.equals(params.get(StartupParams.PropertyKey.USERNAME)))
+                    .findFirst()
+                    .orElse(null);
+            if (user == null) {
+                System.err.println("Couldn't find the provided user in the saved logins. Make sure the username and master password are correct.");
+            }
+        }
+
+        LoginData loginData = new LoginData();
+        loginData.setCredentials(params.get(StartupParams.PropertyKey.USERNAME),
+                (password == null || password.isEmpty()) && user != null ? user.p : password);
```

Comment:
```suggestion
        String username = params.get(StartupParams.PropertyKey.USERNAME),
                masterPw = params.get(StartupParams.PropertyKey.MASTER_PASSWORD),
                password = params.get(StartupParams.PropertyKey.PASSWORD);
        if (username == null || (masterPw == null && (password == null || password.isEmpty()))) {
            System.err.prirntln("Credentials file requires username and either a password or a master password");
            System.exit(-1);
        }
        if (password == null || password.isEmpty()) {
            Credentials credentials = loadCredentials();
            try {
                char[] masterPassword = params.getMasterPassword();
                if (masterPassword == null) {
                    System.err.println("Both master-password and password are undefined, make sure to at least define one of these fields");
                }
                credentials.decrypt(masterPassword);
            } catch (Exception e) {
                System.err.println("Couldn't retreive logins, check your startup properties file");
                e.printStackTrace();
            }
            
            Credentials.User user = credentials.getUsers()
                    .stream()
                    .filter(usr -> username.equals(usr.u))
                    .findFirst()
                    .orElse(null);
            if (user == null) {
                System.err.println("Couldn't find the provided user in the saved logins. Make sure the username and master password are correct.");
                System.exit(-1);
            } else {
                password = user.p;
            }
        }

        LoginData loginData = new LoginData();
        loginData.setCredentials(username, password);
```

I'd consider moving to a different method that is `private String getPassword(String user, String masterPassword)`, returns null if user not found (one with no exits!). It would simplify this function quite a bit